### PR TITLE
Migrate cli-test to Rust

### LINKED ISSUE #2101: Rewriting Scripts in Rust tracking issue
We have a bunch of integration tests written in bash. They are flaky, hard to write and update. We want to re-write all these tests in Rust and minimize the amount of bash we depend on.

We've made 2 PRs in this project so far:
- https://github.com/fedimint/fedimint/pull/1899 Added ability to spawn daemons (e.g. `fedimintd` and `lnd` etc) via Rust
- https://github.com/fedimint/fedimint/pull/1910 Can write tests purely in rust. As a proof-of-concept, it re-wrote latency-test.sh: 
  - [Old bash version that was replaced](https://github.com/fedimint/fedimint/blob/5ef8158163c8cfa120b235e24c07999724bc33a0/scripts/latency-test.sh) 
  - [New rust version](https://github.com/fedimint/fedimint/blob/c77e5960255843890da809b3f3ebdcccc47e27e1/fedimint-bin-tests/src/main.rs#L417-L515) which is called [here](https://github.com/fedimint/fedimint/blob/c77e5960255843890da809b3f3ebdcccc47e27e1/scripts/latency-test.sh)

Work remaining (in order of importance):
- [ ] (1) Export `$FM_LOGS_DIR` in github actions. This helps us debug logs of specific daemons. See [upload-artifact](https://github.com/actions/upload-artifact)
- [x] Rewrite the rest of the bash integrations tests:
  - [x] (2a) [scripts/cli-test.sh](https://github.com/fedimint/fedimint/blob/master/scripts/cli-test.sh) (easier)
  - [x] (2b) [reconnect tests](https://github.com/fedimint/fedimint/blob/master/scripts/reconnect-test.sh) (harder b/c need to stop and start fedimintd)
  - [x] (2c) [upgrade tests](https://github.com/fedimint/fedimint/blob/master/scripts/upgrade-test.sh) (harder b/c need to stop and start fedimintd, but 2b has now demonstrated how to do it)
- [x] (3) Replace [scripts/setup-tests.sh](https://github.com/fedimint/fedimint/blob/master/scripts/setup-tests.sh) with a rust-spawned dev federation. For example, something like [this function](https://github.com/fedimint/fedimint/blob/57c55563680d40080c776cfa0f72b3ec7449255f/fedimint-bin-tests/src/main.rs#L557-L570). Perhaps rename to dev_federation or something. blocked on (2b) (2c)
- [x] (4) Spawn `esplora` in fedimint-bin-tests. This functionality exists in [fedimint-testing/src/bin/fixtures.rs](https://github.com/fedimint/fedimint/blob/89f83f3eb07dbdac21b6eba89f7f5764fcbe6212/fedimint-testing/src/bin/fixtures.rs#L243-L266)
- [x] (5) Spawn `electrs` in fedimint-bin-tests. This functionality exists in [fedimint-testing/src/bin/fixtures.rs](https://github.com/fedimint/fedimint/blob/89f83f3eb07dbdac21b6eba89f7f5764fcbe6212/fedimint-testing/src/bin/fixtures.rs#L224-L241)
- [x] (6) Don't set any environment variables in [scripts/build.sh](https://github.com/fedimint/fedimint/blob/master/scripts/build.sh). Just run `cargo build ...`. Instead, set environment variables in [dev_fed](https://github.com/fedimint/fedimint/blob/57c55563680d40080c776cfa0f72b3ec7449255f/fedimint-bin-tests/src/main.rs#L526-L555) Rust function.
- [x] (7) Once nothing depends on it, remove [scripts/lib.sh](https://github.com/fedimint/fedimint/blob/master/scripts/lib.sh)
- [x] (8) Once nothing depends on it, remove [fedimint-testing/src/bin/fixtures.rs](https://github.com/fedimint/fedimint/blob/master/fedimint-testing/src/bin/fixtures.rs)
- [x] (9) https://github.com/fedimint/fedimint/issues/2137
- [ ] (10) Ability to run dkg with n peers and run `fedimintd` with fewer than n peers 
- [x] (11) Ability to run `fedimintd` without running DKG beforehand. This will allow us to replace [scripts/run-ui.sh](https://github.com/fedimint/fedimint/blob/master/scripts/run-ui.sh) which is useful to test setup UI
- [x] Use `PATH` instead of `FM_BIN_DIR`
- [x] When any `cmd!` fails, log the the entire command that was run
- [x] devimint should peg-in user and gateways for mprocs and tmux developer environments. [Stop using pegin.sh](https://github.com/fedimint/fedimint/blob/a8b6b78328081f000794b5020f6b93c60a8ee7ef/scripts/mprocs-user-shell.sh#L19-L26) and then delete it.
- [x] Write `distributedgen` to a log file. Probably the same one that `fedimintd` uses. This way errors will show up visually in `mprocs`.
- [ ] When `devimint tmuxinator` is running, we should have a command to start / stop the fedimintd servers.
- [x] `include!` [daemon config files](https://github.com/fedimint/fedimint/tree/master/misc/test) so external projects don't need them.
  - [ ] Eventually an option to override these default configs

If you'd like to tackle one of these issues, just post "I'll try (\<n\>)". If multiple people want to try the same thing, try to collaborate in discord. Make [draft pull requests](https://egghead.io/lessons/github-open-a-github-draft-pull-request) early for feedback. @Maan2003 can help you!

### PR DESCRIPTION
**Pull request by GetPsyched**
Progress in #2101 (2a)

### CODE REVIEW

File: fedimint-bin-tests/src/main.rs [hunk 1]
```diff
@@ -516,6 +516,64 @@ async fn latency_tests(dev_fed: DevFed) -> Result<()> {
...
+        "--out-file",
+        "{data_dir}/server-0/config-2"
+    )
+    .env("FM_PASSWORD", "pass-foo");
+
+    cmd!(
+        "{bin_dir}/distributedgen",
+        "config-decrypt",
+        "--in-file",
+        "{data_dir}/server-0/config-2",
```
Comment (Maan2003) [inline fedimint-bin-tests/src/main.rs:561]: you can use `"--in-file={data_dir}/...."`

looks cleaner

File: fedimint-bin-tests/src/util.rs [hunk 2]
```diff
@@ -69,6 +70,15 @@ impl Command {
         self
     }
 
+    pub fn env<K, V>(mut self, key: K, val: V) -> Self
+    where
+        K: AsRef<OsStr>,
+        V: AsRef<OsStr>,
```
Comment (Maan2003) [inline fedimint-bin-tests/src/util.rs:76]: nice, please also add these to `envs`

**Comment by justinmoon** [comment]
Nice progress

**Comment by GetPsyched** [comment]
Thanks, I'll keep an eye out at #2058 and make sure to reflect the changes in the migrated script.

**Comment by GetPsyched** [comment]
I am unable to run the test at local due to 2 errors in `main.rs` @ [L333](https://github.com/fedimint/fedimint/blob/9894c95f512e5f086b6358480112dadc2d2f1844/fedimint-bin-tests/src/main.rs#L333) and [L366](https://github.com/fedimint/fedimint/blob/9894c95f512e5f086b6358480112dadc2d2f1844/fedimint-bin-tests/src/main.rs#L366) which I have not changed in this commit.

File: fedimint-bin-tests/src/main.rs [hunk 3]
```diff
@@ -563,21 +565,368 @@ async fn cli_tests(dev_fed: DevFed) -> Result<()> {
...
     );
 
-    fed.pegin(10_000).await?;
-    // fed.pegin_gateway(99_999).await?;
+    fed.pegin(10_000).await;
+    fed.pegin_gateway(99_999, &gw_cln).await;
 
-    let connect_string = fs::read_to_string(format!("{data_dir}/client-connect")).await?;
-    fs::remove_file(format!("{data_dir}/client.json"));
+    let connect_string = fs::read_to_string(format!("{}/client-connect", data_dir)).await?;
```
Comment (dpc) [inline fedimint-bin-tests/src/main.rs:622]: I thought that `clippy` nags now if you don't use inline formatting form when poossible.

File: fedimint-bin-tests/src/main.rs [hunk 4]
```diff
@@ -563,21 +565,368 @@ async fn cli_tests(dev_fed: DevFed) -> Result<()> {
...
     assert_eq!(
-        cmd!(fed, "decode-connect-info", connect_string)
+        cmd!(fed, "decode-connect-info", connect_string.clone())
             .out_json()
             .await?["id"]
             .as_str(),
         Some(&*fed_id),
     );
+    // Number required for one honest is ceil(($FM_FED_SIZE-1)/3+1)
+    let one_honest = 2;
```
Comment (dpc) [inline fedimint-bin-tests/src/main.rs:635]: `one_honest_threashold`. I never like that "one honest" as integer thing. It's a bit `one = 2`, to me.

File: fedimint-bin-tests/src/main.rs [hunk 5]
```diff
@@ -563,21 +565,368 @@ async fn cli_tests(dev_fed: DevFed) -> Result<()> {
...
+    let client_json_str = fs::read_to_string(format!("{}/client.json", data_dir)).await?;
+    let client_json: serde_json::Value = serde_json::from_str(&client_json_str)?;
+    let api_endpoints = client_json["api_endpoints"].as_object().unwrap();
+    let one_honest_urls: Vec<&str> = api_endpoints
+        .iter()
+        .take(one_honest)
```
Comment (dpc) [inline fedimint-bin-tests/src/main.rs:641]: Here `take(one_...)` means "take two" :D

File: fedimint-bin-tests/src/main.rs (hunk 4 above)
Comment (dpc) [reply fedimint-bin-tests/src/main.rs:635]: @jkitman You're probably the best person to talk about it. You see my point (here and below)?

**Review by dpc — COMMENTED** [review]
Made some minor comments, but generally LGTM.

File: fedimint-bin-tests/src/federation.rs [hunk 6]
```diff
@@ -105,6 +105,34 @@ impl Federation {
...
+        cmd!(
+            gw_cln,
+            "deposit",
+            "--federation-id={fed_id}",
+            "txout-proof={txout_proof}",
+            "--raw-tx={raw_tx}"
+        )
+        .run()
+        .await?;
+        cmd!(self, "fetch").run().await?;
```
[the remark this answers is not in the document]
Comment (dpc) [reply fedimint-bin-tests/src/federation.rs:162]: OK then. Seems much nicer than then what it expands to so probably worth it.

**Comment by justinmoon** [comment]
> I am unable to run `cli-test.sh` due to 2 errors in `main.rs` @ [L333](https://github.com/fedimint/fedimint/blob/9894c95f512e5f086b6358480112dadc2d2f1844/fedimint-bin-tests/src/main.rs#L333) and [L366](https://github.com/fedimint/fedimint/blob/9894c95f512e5f086b6358480112dadc2d2f1844/fedimint-bin-tests/src/main.rs#L366) which have not been changed in this PR.

If you change line 333 to this, it runs:

```
let gateway_env: HashMap<String, String> = match ln {
```

File: scripts/cli-test.sh [hunk 7]
```diff
@@ -3,114 +3,6 @@
...
-# CLN can pay LND directly
-ADD_INVOICE="$($FM_LNCLI addinvoice --amt_msat 42000)"
-INVOICE="$(echo $ADD_INVOICE| jq -e -r '.payment_request')"
-PAYMENT_HASH="$(echo $ADD_INVOICE| jq -e -r '.r_hash')"
-$FM_LIGHTNING_CLI pay "$INVOICE"
-INVOICE_STATUS="$($FM_LNCLI lookupinvoice $PAYMENT_HASH | jq -e -r '.state')"
-[[ "$INVOICE_STATUS" = "SETTLED" ]]
-
-echo "fm success: cli-test"
+$FM_BIN_DIR/fedimint-bin-tests latency-tests
```
Comment (justinmoon) [inline scripts/cli-test.sh:8]: Should be `$FM_BIN_DIR/fedimint-bin-tests cli-test`

File: fedimint-bin-tests/src/main.rs (hunk 4 above)
Comment (jkitman) [reply fedimint-bin-tests/src/main.rs:635]: I'm changing the connect info to be one per guardian in order to accommodate the join tokens.  Would like to get the changes in https://github.com/fedimint/fedimint/pull/2175 merged before this PR.

Comment (jkitman) [reply fedimint-bin-tests/src/main.rs:589]: @GetPsyched Sorry for breaking these tests, if you look at the `cli-test.sh` in master you will see how to fix it.

File: fedimint-bin-tests/src/main.rs [hunk 8]
```diff
@@ -567,6 +569,419 @@ async fn reconnect_test(dev_fed: DevFed, process_mgr: &ProcessManager) -> Result
...
+        fed,
+        "ln-invoice",
+        "--amount=100000msat",
+        "--description='incoming-over-lnd-gw'"
+    )
+    .out_json()
+    .await?["invoice"]
+        .as_str()
+        .context("invoice must be string")?
+        .to_owned();
```
[the remark this answers is not in the document]
Comment (GetPsyched) [reply fedimint-bin-tests/src/main.rs:731]: I think it's by the `rust-analyzer` VSC extension

File: scripts/cli-test.sh (hunk 7 above)
Comment (GetPsyched) [reply scripts/cli-test.sh:8]: Oh no, I blame fatigue!

File: fedimint-bin-tests/src/main.rs [hunk 9]
```diff
@@ -521,6 +523,401 @@ async fn latency_tests(dev_fed: DevFed) -> Result<()> {
...
+        .request(cln_rpc::model::InvoiceRequest {
+            amount_msat: AmountOrAny::Amount(ClnRpcAmount::from_msat(100_000)),
+            description: "lnd-gw-to-cln".to_string(),
+            label: "test".to_string(),
+            expiry: Some(1),
+            fallbacks: None,
+            preimage: None,
+            exposeprivatechannels: None,
+            cltv: None,
+            deschashonly: None,
```
Comment (justinmoon) [inline fedimint-bin-tests/src/main.rs:770]: In any of these places where there a bunch of `None` fields in a `cln_rpc::model::*` struct, you can replace them all with a single `Default::default(),` line. This should make the test a little less verbose.

Comment (Maan2003) [reply fedimint-bin-tests/src/main.rs:770]: `cln_rpc::model::*` don't implement Default :/

Comment (justinmoon) [reply fedimint-bin-tests/src/main.rs:770]: Ohhh ... that's annoying

Comment (justinmoon) [reply fedimint-bin-tests/src/main.rs:770]: Made an issue about this upstream https://github.com/ElementsProject/lightning/issues/6174

Comment (GetPsyched) [reply fedimint-bin-tests/src/main.rs:770]: I get the following error when I try to do this:
> the trait `std::default::Default` is not implemented for `*`
`*` being the `cln_rpc::model` struct I've used

File: fedimint-bin-tests/src/main.rs (hunk 4 above)
Comment (GetPsyched) [reply fedimint-bin-tests/src/main.rs]: No worries, I've added your changes and everything should work fine.

**Review by justinmoon — APPROVED** [review]

**Comment by justinmoon** [comment]
Great work @GetPsyched!

**Comment by justinmoon** [comment]
> Looks good, wonder if we can add these to our code cov tests so we can get a more accurate measurement

@jkitman made an issue https://github.com/fedimint/fedimint/issues/2213

**Pull request merged**